### PR TITLE
Follow up on #8085 - refactor com_contact form to support renderField

### PR DESCRIPTION
**Pull request by matrikular**
**Introduction**
As a follow-up on #8085 I would like to propose a few changes first to com_contact and later on to com_users as well.

_While there are a couple of code style changes in files that are not directly related to com_contact, I thought it would be good to include them in this PR since they were edited in the same context._

**Status quo**
The last PR allows us to override and / or add fields to the contact form via plugins without the need to create a template override because the rendering is based on the xml file. As discussed in the comment section in #8085, we normally want to use the renderField method instead of writing things like echo $field->input. For that, a component specific renderfield override had to be created, which I did with this PR.

The method used to check for a required state and / or field of the type label is not very clean but should suffice for now. I normally would have created or rewritten the fields and their layouts but that would have been too much for this PR and, as @wilsonge told me, that is planned for the near future. Take a look at [layouts/joomla/form/field/radio.php](https://github.com/joomla/joomla-cms/blob/staging/layouts/joomla/form/field/radio.php) to see where this is going.

**Changes**
I've tried to document / describe the changes made to every file within the description of the file change request. Without going too much into the details, let me please give you a short summary:
- Refactor the form layout to make use of the renderField method. In doing so, the legend for each fieldset (if present) will be taken from the xml file and the markup for each field will be based on the renderfield.php and the respective layout. As a bonus, the contact form now supports the [showon-Attribute](https://docs.joomla.org/Form_field#Showon).
- Some language strings, like the fieldset label and the "Send copy to yourself" have been changed.
- The form markup was changed to match com_users.registration and login which includes adding a "well" CSS class suffix to the form.

**Before**
contact_before
**After**
contact_after

My goal was to unify the output / markup, to fix some bugs and add some functionality to all of the forms of com_contact and com_users (registration / login).

**Long story short, how to test?**
Well, first of all, this only works for the latest dev branch. Second, after applying the patch the output should look like in the second picture. Apart from that you can follow the test instructions mentiond in #8085. I've uploaded a new version of the test plugin [here](http://www.curiosys.net/plg_system_contact.zip).

If you have any questions, suggestions or corrections feel free to contact me.

### CODE REVIEW

File: components/com_contact/views/contact/tmpl/default_form.php [hunk 1]
```diff
@@ -9,44 +9,31 @@
...
-	<form id="contact-form" action="<?php echo JRoute::_('index.php'); ?>" method="post" class="form-validate form-horizontal">
-		<fieldset>
-			<legend><?php echo JText::_('COM_CONTACT_FORM_LABEL'); ?></legend>
-
-			<?php foreach ($this->form->getFieldsets() as $fieldset) : ?>
-				<?php foreach ($this->form->getFieldset($fieldset->name) as $field) : ?>
-					<?php if ($field->name === 'contact_email_copy' && !$this->params->get('show_email_copy')) : ?>
-						<?php continue; ?>
+	<form id="contact-form" action="<?php echo JRoute::_('index.php'); ?>" method="post" class="form-validate form-horizontal well">
+		<?php foreach ($this->form->getFieldsets() as $fieldset): ?>
```
Comment (zero-24) [inline components/com_contact/views/contact/tmpl/default_form.php]: Can you add a space between `$fieldset)` and `: ?>`

**Comment by bembelimen** [comment]
I have tested this item :white_check_mark: successfully on e119c7f581b183f5293e61848de5d184fba8c64d<br /><br />Two general comments:

- I saw, the new way to gather variables is "extract" (used in several other layouts/tmpl), in personal I think, that's not a solid way to code (often mentioned in "bad coding practises). Perhaps that function should be reviewed in Joomla! general.
- If we want to check if a string is not empty, we don't need the StringHelper::strlen() method, strlen is enough (also on utf8 strings), perhaps a (very) small performance improvement

Otherwise the PR looks very good. Awesome work!
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8473">issues.joomla.org/joomla-cms/8473</a>.</sub>

**Comment by ufuk-avcu** [comment]
Tested on FF 40 - the output looks like the second screenshot. An empty <fieldset></fieldset> is in the html code, if no Captcha is used.

**Comment by matrikular** [comment]
Thank you for your comments and tests.

The StringHelper dependency was reverted back to use the nativ strlen function and the foreach loop extended to only render the fieldset when captcha was enabled and (enabled) catpcha plugins exists.

It would mean a lot to me if you could test it again, please.

**Comment by chmst** [comment]
Ive tested this item successfully<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8473">issues.joomla.org/joomla-cms/8473</a>.</sub>

**Comment by chmst** [comment]
I have tested this item :white_check_mark: successfully on a16cc179fbb7d2d4f8e8253c70e4d9ff9143c24d<br /><br />successful on 3.5<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8473">issues.joomla.org/joomla-cms/8473</a>.</sub>

**Comment by brianteeman** [comment]
By adding the well class to the form you are potentially changing the display on existing web sites not sure if that passes the B/C promise

**Comment by matrikular** [comment]
@brianteeman thank you for your input.

I would like to see the addition of the well class suffix to the form more of a bug fix than a potential b/c issue. If we look at the registration and login view of com_users e.g, both forms have the well suffix. Adding it would add to the general user experience (forms look like "this" ~> POLS).

I've checked approx. 50 sites (google / j.org showcase) where this would have no negative impact.

**Comment by brianteeman** [comment]
It would change the layout on all sites that I have built. Not sure how happy I am about that.

**Comment by Bakual** [comment]
Templates aren't covered by our B/C policy. So this PR doesn't **have** to be backward compatible in this regard.
Adding the well sure will not break any template functionality. It will obviously change appearance on most templates. The real question in the end thus is if it is considered an improvement or not and if it's worth it.

**Comment by chmst** [comment]
My2Cent: I always make overrides for the contact form... because it is ugly as it is, with the well only around the button. It is an improvement :)

**Comment by kolvar** [comment]
Tested it as described. Works as expected. Fine work.

**Comment by matrikular** [comment]
As far as I can tell, it's not this PR that fails, correct? (~PHP7 build)

**Comment by Bakual** [comment]
Yeah, it's an issue related to a hardware outage on pear.net. See https://github.com/travis-ci/travis-ci/issues/5207

**Comment by wilsonge** [comment]
Tests pass in the non-PHP 7 libraries and there are two good tests here. merging

**Comment by infograf768** [comment]
@wilsonge
Please change Milestone as this is now in staging...

**Comment by infograf768** [comment]
Also, a string has been deleted
`COM_CONTACT_FORM_LABEL="Send an Email. All fields with an asterisk (*) are required.`
This is not B/C as the language packs can be used on pre-3.5.0 installs.
I am making a PR for that.

**Comment by infograf768** [comment]
Please merge on review https://github.com/joomla/joomla-cms/pull/8602 (also correcting alpha order)
